### ISSUE TITLE
GS can't eat its own dog food.

### ISSUE THREAD

**Issue by iham** [plone/Products.CMFPlone#1794]
**situation**
set up a plone 5.1a2 site.
create a snapshot in portal_setup.
change the day the week should start with in the control panel "Date and Time Settings".
create a snapshot in portal_setup again and make a diff between those two.

i then safed this changed settings to my registry.xml

```
  <record name="plone.first_weekday" interface="Products.CMFPlone.interfaces.controlpanel.IDateAndTimeSchema" field="first_weekday">
    <field type="plone.registry.field.Choice">
      <description xmlns:ns0="http://xml.zope.org/namespaces/i18n" ns0:domain="plone" ns0:translate="help_first_weekday">First day in the week.</description>
      <title xmlns:ns0="http://xml.zope.org/namespaces/i18n" ns0:domain="plone" ns0:translate="label_first_weekday">First weekday</title>
      <vocabulary>plone.app.vocabularies.Weekdays</vocabulary>
    </field>
    <value>3</value>
  </record>
```

now i set up another plone site and also install my package with this settings stored inside and get an error:

```
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module Products.CMFPlone.browser.admin, line 272, in __call__
  Module Products.CMFPlone.factory, line 168, in addPloneSite
  Module Products.GenericSetup.tool, line 379, in runAllImportStepsFromProfile
   - __traceback_info__: profile-akbild.site:default
  Module Products.GenericSetup.tool, line 1414, in _runImportStepsFromContext
  Module Products.GenericSetup.tool, line 1226, in _doRunImportStep
   - __traceback_info__: plone.app.registry
  Module plone.app.registry.exportimport.handler, line 47, in importRegistry
  Module plone.app.registry.exportimport.handler, line 90, in importDocument
  Module plone.app.registry.exportimport.handler, line 246, in importRecord
   - __traceback_info__: record name: plone.first_weekday
  Module plone.supermodel.utils, line 144, in elementToValue
  Module zope.schema._field, line 322, in fromUnicode
  Module zope.schema._bootstrapfields, line 183, in validate
  Module zope.schema._field, line 338, in _validate
ConstraintNotSatisfied: (u'0', '')
```

0 -> Monday (Europe), but every other number from 0 (monday) to 6 (sunday) generates the same error.

**why does that happen?**
the choice field (zope.schema) validates a given value (the one from the xml file -> u'0') against its vocabulary, using by_value, which is - in case of plone.app.vocabularies.datetimerelated.WeekdaysFactory - a dict of number (keys) and simpleterm (value) - pairs. but the value from the xml is a unicode string, not an int. -> so its not a key in the vocabulary.  see zope.schema._field.Choice._validate() for that.

``` python
u'0' in {0:term_a, 1:term_b} == False
```

**further down the rabbit hole...**
most SimpleTerms in plone.app.vocabularies are instantiated by a value (string) and a title (term); only the Month and Weekday Vocabularies are build different. the come with a value, token and title. value is an index, token a string representation of that index and the title (term). this is, at least, inconsistent. 

**solutions**
1. rewrite the vocabularies to have better key/value consistency and fix cmfplone Products.CMFPlone.setuphandlers.first_weekday_setup to use strings instead of integers.
2. extend the choice field to also handle value_type; like IDict and ICollection does.
3. make all GS setup files JSON instead of XML (no... just kidding... but would be awesome)

i will implement the first solution, as i am not able (too heavy) to extend the Choice field or reimplement the whole GS-setup infrastructure.

**Comment by seanupton** [comment]
Is there a reason we need to validate inputs here?  IMHO, zope.schema._field.Choice.fromUnicode is fundamentally broken for non-string values, and as such we should bypass validating containment on any Choice field where vocabulary is numeric, period.  A way to do this is proposed [here](https://github.com/plone/Products.CMFPlone/issues/1268#issuecomment-282826666) -- though in that case permissive vocabulary checks are proposed for an entirely different setup profile issue/reason (chicken/egg problem with portal_type names).

I would propose that this is sensible for *all* index-keyed vocabularies in datetimerelated.py -- and may work on that along-side my fix for #1268.

_FWIW, I'm not sure if this same issue is going to be tricky with tuple-based values for image scales vocabulary, so I would punt on even touching that._

**Comment by seanupton** [comment]
Folks, I have a pull request open in plone.app.vocabularies that I believe will fix this.  See above.

**Comment by seanupton** [comment]
I think this can be closed?  My fix referenced above by @mister-roboto (merged) fixes this and also #1268.
